fish-rust/src/builtins/random.rs

```diff
@@ -0,0 +1,209 @@
+    let mut start = 0;
+    let mut end = 32767;
+    let mut step = 1;
+    let arg_count = argc - w.woptind;
+    let i = w.woptind;
+    if arg_count >= 1 && argv[i] == "choice" {
+        if arg_count == 1 {
+            streams.err.append(wgettext_fmt!(
+                "%ls: nothing to choose from\n",
+                cmd,
+            ));
+            return STATUS_INVALID_ARGS;
+        }
+
+        let rand = engine.gen_range(0..arg_count - 1);
+        streams.out.append(format::printf::sprintf!("%ls\n"L, argv[i + 1 + rand]));
+        return STATUS_CMD_OK;
+    }
+    let mut parse_ll = |num : &wstr| {
+        let res: Result<i64, wutil::Error> = fish_wcstoi_radix_all(num.chars(), None, true);
+        if res.is_err() || res.unwrap() < 0 {
+            match res {
+                Err(_) => {
+                    streams.err.append(wgettext_fmt!(
+                        "%ls: %ls: invalid integer\n",
+                        cmd,
+                        num,
+                    ));
+                },
+                Ok(_) => {}
+            }
+        }
+        return res;
+    };
+
+    match arg_count {
+        0 => {
+            // Keep the defaults
+        },
+        1 => {
+            // Seed the engine persistently
+            let num = parse_ll(argv[i]);
+            if num.is_err() {
+                return STATUS_INVALID_ARGS;
+            }
+            *engine = SmallRng::seed_from_u64(num.unwrap() as u64);
+            return STATUS_CMD_OK;
+        },
+        2 => {
+            // start is first, end is second
+            let num = parse_ll(argv[i]);
+            if num.is_err() {
+                return STATUS_INVALID_ARGS;
+            }
+            start = num.unwrap();
+
+            let num = parse_ll(argv[i + 1]);
+            if num.is_err() {
+                return STATUS_INVALID_ARGS;
+            }
+            end = num.unwrap();
+        },
+        3 => {
+            // start, step, end
+            let num = parse_ll(argv[i]);
+            if num.is_err() {
+                return STATUS_INVALID_ARGS;
+            }
+            start = num.unwrap();
+
+            let num = parse_ll(argv[i + 1]);
+            if num.is_err() {
+                return STATUS_INVALID_ARGS;
+            }
+            if num.unwrap() < 0 {
+                // XXX: Historical, this read an "unsigned long long"
+                // This should actually be "must be a positive integer"
+                streams.err.append(wgettext_fmt!(
+                    "%ls: %ls: invalid integer\n",
+                    cmd,
+                    argv[i + 1],
+                ));
+                return STATUS_INVALID_ARGS;
+            }
+            if num.unwrap() == 0 {
+                streams.err.append(wgettext_fmt!(
+                    "%ls: STEP must be a positive integer\n",
+                    cmd,
+                ));
+                return STATUS_INVALID_ARGS;
+            }
+            step = num.unwrap();
```

Comment:
You have a lot of calls to `.unwrap()` but they're all cases where you can avoid that altogether by using match blocks. That's better because you never want to get into the habit of glossing over `.unwrap()` because you use it liberally.

You can either do 

```let num = match num {
    Ok(num) => num,
    Err(e) => { streams.err.append(...); return STATUS_XXX; }
};

// then continue to just use `num` instead of `num.unwrap()` thereafter
```

or you can convert the entire highlighted block into one match block, which is probably more idiomatic because it lets the compiler do things like check for untested branches, check for duplicate branches, check for branches that can't be taken because of a previous check, etc.

```rust
let step = match num {
    Err(e) => // as before,
    Ok(x) if x < 0 => ..., // or, more compactly, Ok(..0) => ...
    Ok(0) => ....,
    Ok(x) => x,
};
```

Reply:
Okay. I left the one for the rng because that's something we need to look at regarding concurrent execution.